Updates springer article

eliocamp opened the pull request:
Closes #491

cderv commented on inst/rmarkdown/templates/springer/skeleton/sn-apacite.bst:
```diff
@@ -0,0 +1,5357 @@
+%%
```
Is this file still us somewhere ?

eliocamp replied:
What do you mean?

cderv replied:
Sorry for the typo.  I meant

"Is this file still use somewhere ?"

cderv replied:
And now I understand that it is required by the classoption. All good !

cderv commented on R/article.R:
```diff
@@ -490,7 +490,7 @@ sim_article <- function(..., highlight = NULL, citation_package = "natbib") {
 springer_article <- function(..., keep_tex = TRUE,
-                             citation_package = "default") {
+                             citation_package = "natbib") {
   pdf_document_format(
```
We need to add a Pandoc requirement in there as we are using some recent feature like `$it.` in template. 

You can find such R line of code in some other template

eliocamp replied:
Do you know which version is required?

eliocamp replied:
Well, GitHub Actions failed for low pandoc versions. I added 

```r
  if (!rmarkdown::pandoc_available("2.11.4")) {
    stop("`springer_article()` now requires a minimum of pandoc 2.11.4")
  }
```

Because 2.11.4 is the oldest version that didn't fail.

cderv reviewed (commented):
Sorry for the delay on this one. And thanks a lot for the work !!

Several other comments in additions to the inlines ones: 

* The documentation of the function should probably be updated
* Are the `.bst` file used somewhere ? The line in the TeX template for using it are removed I think
* Some Pandoc's requirement in TeX template are missing and some Pandoc's feature won't probably work (like table, highlighting, etc...) I think we should add them.  

Did you use the format for your submission already ? Was it accepted ? 

And please, share with me the status. If you can't work on that more, I need to know to organize and take care of it. If you can but need time, no worries we are not in a hurry. So just asking for organization

cderv commented:
@eliocamp are you planning to come back to this ? 

If not, I'll try to update myself. Just need to know as it's been some time now.

eliocamp commented:
Thanks for the reminder! I'll get to it this week. Sorry!

eliocamp commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -1,29 +1,74 @@
+%%\documentclass[default]{sn-jnl}% Default
+%%\documentclass[default,iicol]{sn-jnl}% Default with double column layout
+
+%%%% Standard Packages
+%%<additional latex packages if required can be included here>
+%%%%
```
Not really. Added packages in hearder-includes and it was fine.

eliocamp commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -90,82 +135,62 @@
-$if(titlerunning)$
-    \titlerunning{$titlerunning$}
-$endif$
+\title[$if(titlerunning)$$titlerunning$$endif$]{$title$}
```
They are not used in the template; I guess not.

eliocamp commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
````diff
@@ -1,63 +1,380 @@
-# Introduction {#intro}
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = FALSE,
+                      out.extra = "")  # This forces knitr to label all figures.
````
It's kind of a workaround. Knitr will not label figures that have no caption; adding `out.extra` to be anything forces knitr to add labels. It's not required but it can be confusing if its not there.  It's on the AGU template too.

cderv replied:
Oh interesting. This is indeed a required workaround to get **knitr** to produce LaTeX code directly and not let pandoc handles image. 

Probably something we can improve in **knitr** part of https://github.com/yihui/knitr/issues/1991

Let's leave it then. Thanks for the explanation

eliocamp commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
```diff
@@ -1,63 +1,380 @@
+\end{lstlisting}
+\end{minipage}
+
+# Cross referencing {#sec8}
+
+Figures and tables are labeled with a prefix (fig or tab, respectively) plus the chunk label. 
+Other environments such as equation and align can be labelled via the \verb+\label{#label}+ command inside or just below the \verb+\caption{}+ command.
+You can then use the label for cross-reference. As an example, consider
+the chunk label declared for Figure\ \ref{fig:fig1} which is
+fig1. 
+To cross-reference it, use the command
+\verb+Figure \ref{fig:fig1}+, for which it comes up as
+``Figure\ \ref{fig:fig1}''.
+
+To reference line numbers in an algorithm, consider the label declared for the line number 2 of Algorithm\ \ref{algo1} is \verb+\label{algln2}+. To cross-reference it, use the command \verb+\ref{algln2}+ for which it comes up as line\ \ref{algln2} of Algorithm\ \ref{algo1}.
```
The raw LaTeX should work by default. In my version I used `bookdown`  and `\@ref(fig:figure)` to render the cross-references in other formats.

eliocamp commented:
Well, hell. They updated the template last month :zany_face: . Now I updated it to the current version. And added a lot of your suggestions. 


> The documentation of the function should probably be updated

Do you mean the function documentation for `springer_article()`? What should I add?

> Are the .bst file used somewhere ? The line in the TeX template for using it are removed I think

The bst files are needed for the class options (e.g. classoption sn-nature requires sn-nature.bst). They come with the template you download from the publisher. 

>  Some Pandoc's requirement in TeX template are missing and some Pandoc's feature won't probably work (like table, highlighting, etc...) I think we should add them.

Which requirements are missing?

eliocamp commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -1,29 +1,74 @@
+
+\jyear{$year$}%
+
+%% as per the requirement new theorem styles can be included as shown below
+\theoremstyle{thmstyleone}%
+\newtheorem{theorem}{Theorem}%  meant for continuous numbers
+%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
+%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
+\newtheorem{proposition}[theorem]{Proposition}%
+%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.
+
+\theoremstyle{thmstyletwo}%
+\newtheorem{example}{Example}%
+\newtheorem{remark}{Remark}%
+
+\theoremstyle{thmstylethree}%
+\newtheorem{definition}{Definition}%
```
I think this could be in `header-includes`?

cderv replied:
If this is not too late to define it, then indeed it is a good place ! All good then.

eliocamp commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
```diff
@@ -1,63 +1,380 @@
+  **Conclusion:** The abstract serves both as a general introduction to the topic and as a brief, non-technical summary of the main results and their implications. The abstract must not include subheadings (unless expressly permitted in the journal's Instructions to Authors), equations or citations. As a guide the abstract should not exceed 200 words. Most journals do not set a hard limit however ...
 
 bibliography: bibliography.bib
-biblio-style: spphys
-# bibstyle options spbasic(default), spphys, spmpsci
```
I understand that the reference styles are encapsulated in the various classoptions, like sn-basic, sn-nature, usw...

cderv replied:
Oh ok ! Every tools do its thing differently so I am easily lost between all the templates. Thanks

eliocamp commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
```diff
@@ -1,63 +1,380 @@
+
+\backmatter
+
+\bmhead{Supplementary information}
```
I don't think so. [Someone is having a similar problem](https://stackoverflow.com/questions/75321227/adding-classes-to-headers-with-a-pandoc-lua-filter) trying to adapt this template to quarto and their solution is creating a custom lua filtre, which might be overkill here (and I don't know how to do it :P).

cderv replied:
Yeah we could also do it in rticles with Lua filter but I agree that this could be too much. 

For Quarto, there is already a `latex-environment` extension that handles command that could be useful for this

Let's leave it

eliocamp commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -1,29 +1,74 @@
-% Insert the name of "your journal" with
-% \journalname{myjournal}
-%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
What did we miss?

cderv replied:
Not specfic to what is already included or not but Pandoc's feature are 

* LaTeX part CSL handling for Pandoc's citeproc 
* LaTeX part for table handling
* LaTeX part for graphic handling
* LaTeX part for footnotes, verbatim in footnotes, link as not and other feature...

Not exhaustive. 

I'll add what I think is necessary as other template (tables, graphics and CSL mainly)

cderv replied:
I see in the template 

* highlighting-macros
* tables
* CSL refs
* tightlist

So I think overall we are ok. 

Let's leave it that way and see later if some content will have issues.

cderv commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
```diff
@@ -1,63 +1,481 @@
+
+classoptions: 
+  - sn-basic      
+  - pdflatex      # Optional: Compile with pdflatex/xelatex
```
This is a case we had in another template, and just to share that it can be done this way to automate the value based on the usual `latex_engine` configuration in the YAML 

https://github.com/rstudio/rticles/blob/ba7bb2e052bea1fdd28a8bfbb6723fde0f3307ec/R/article.R#L416-L419

https://github.com/rstudio/rticles/blob/ba7bb2e052bea1fdd28a8bfbb6723fde0f3307ec/inst/rmarkdown/templates/mdpi/resources/template.tex#L7

(note that we are missing classoption above)

Anyway, not simpler maybe, just more automated based on other configuration. Because if `latex_engine` is set to `pdflatex` (the default), or to `xelatex` this option will be required. If set to `lualatex` it needs to be removed probably. 

Just sharing in case you find that useful.

cderv commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -1,32 +1,77 @@
+
+\raggedbottom
+
+$if(unnumbered)$
+\unnumbered uncomment this for unnumbered level heads
+$endif$
```
We could also make this variable set based on `number_sections` configuration value in `springer_format()` as this is one of the existing argument in `pdf_document()`

`mdpi_article()` has example of what can be done to set variables for template

cderv commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
```diff
@@ -1,63 +1,481 @@
+  # - lineno        # Optional: Add line numbers
+  # - iicol         # Optional: Double column layour
+
+# unnumbered: TRUE # For unnumbered level heads
```
We have the `number_sections` in `pdf_document()` which is FALSE by default. 

I understand there is no conflict by unsetting this variable ? I would have expected that 
```
unnumbered: true
```
would be the default in that case... 

Maybe we should set `number_sections = TRUE` in the R function then ?

Or did you observe no effect at all ?

cderv commented on inst/rmarkdown/templates/springer/skeleton/skeleton.Rmd:
I wonder if we should try to warn user that the template has changed and previous Rmd file won't work ... 

By removing some variables (`authorrunning`, `thanks`), modifying some like `authors`, and requiring others like `classoption`

This could be done in some way by 

* Adding a `pre_knit` step in the format
* Checking if `metadata` is containing some old values or not containing some expected value

Example: 

* We could check that `classoption` is indeed present and if not warn about the new template and new skeleton
* We could check each fields that we know could be problematic

Could be too much if we consider the work to do regarding the benefit. Maybe adding in NEWS.md has breaking change is enough. 

We could also come up with a new solution and add a mechanism to message user once per session that the format function has a new template in this new version of **rticles**

What do you think about all this and how to handle this breaking change ?

eliocamp replied:
Yeah, versioning might be an issue for rticles in general. In this case I think checking for the presence of `classoption` might be enough, but it would be nice to have a more "self-contained" workflow.

cderv reviewed (commented):
Thanks a lot for the update. 

I have new comments following this new template. Thanks for keeping up with me ! :) 

Overall, this is really good and I am seeking improvment to make the experience less painful for users, especially because we are breaking stuff. 

Please share thoughts, and I am happy to finish the work if you can't or don't see what I mean. 

Waiting for your feedback. Thank you.

eliocamp commented on inst/rmarkdown/templates/springer/resources/template.tex:
```diff
@@ -1,32 +1,77 @@
+%% Please do not use \input{...} to include other tex files.       %%
+%% Submit your LaTeX manuscript as one .tex document.              %%
+%%                                                                 %%
+%% All additional figures and files should be attached             %%
+%% separately and not embedded in the \TeX\ document itself.       %%
```
I have no idea, honestly.  In general use, rticles would create a single .tex file so by default it seems to comply with this order, so I'd say that we don't need to worry about it.

cderv commented:
@eliocamp do you have time to look at it today ? I would like to do a release for tomorrow or Monday at last. It would be good to have this new formats ! Sorry for pushing on this. If you can't, I'll handle myself and make the choices.

eliocamp commented:
I can work on it tomorrow.

eliocamp commented on R/article.R:
```diff
@@ -489,11 +489,15 @@ sim_article <- function(..., highlight = NULL, citation_package = "natbib") {
   pdf_document_format(
     "springer",
-    keep_tex = keep_tex, citation_package = citation_package, ...
+    keep_tex = keep_tex, citation_package = "natbib", ...
```
> Does this template only requires natbib for sure ?

No, I'm not sure, but the instructions in the template are tu use `\citet`. I can just use the default and assume that anyone willing to change it would understand what they're doing.

eliocamp commented:
I think that's all.

cderv reviewed (approved):
Awesome thanks ! 

I believe there is still a few adjustment, but I'll tweak slightly before merging. thank you!

cderv commented:
@eliocamp did you check the theorem style environment are working ?

I  get this warning
````
Warning: Package amsthm Warning: Unknown theoremstyle `thmstyleone' on input line 49.
Warning: Package amsthm Warning: Unknown theoremstyle `thmstyletwo' on input line 56.
Warning: Package amsthm Warning: Unknown theoremstyle `thmstylethree' on input line 60.
```

and it seems the default styles defined in the `sn-jnl.cls` is not working. Did you notice that too ?

eliocamp commented:
Yeah, I got those too but I don't understand what the warning is about. Thet come up even with the unmodified template.

cderv commented:
OK it seems the class file is broken on this then. Maybe not the right way to create style environment. 

I'll move the theorem part as header include then to make it optional and customizable and not hard code those definitions. thanks !
